extensions/notebook/src/book/remoteBookController.ts

```diff
@@ -0,0 +1,290 @@
+			// Get workspace root path
+			let folders = vscode.workspace.workspaceFolders;
+			this._local_path = new URL(folders[0].uri.fsPath);
+		} else {
+			//If no workspace folder is opened then path is Users directory
+			this._local_path = new URL(utils.getUserHome());
+		}
+	}
+}
+
+class SharedRemoteBook extends RemoteBook {
+	constructor(public remote_path: URL) {
+		super(remote_path);
+	}
+
+	// TODO: Not yet supported
+	public async createLocalCopy(): Promise<void> {
+		throw new Error('Not yet supported');
+	}
+}
+
+export class GitHubRemoteBook extends RemoteBook {
+	readonly re = /\//g;
+
+	constructor(public remote_path: URL, protected _assetUrl: URL) {
+		super(remote_path, _assetUrl);
+	}
+
+	public async createLocalCopy(): Promise<void> {
+		this.outputChannel.show(true);
+		this.outputChannel.appendLine(msgRemoteBookDownloadProgress);
+		this.setLocalPath();
+		let fileName = this._assetUrl.pathname.substring(1).replace(this.re, '-').concat('-', utils.generateGuid());
+		this._local_path = new URL(path.join(this._local_path.href, fileName));
+
+		return new Promise((resolve, reject) => {
+			fs.mkdir(this._local_path.href, (err) => {
+				if (err) {
+					this.outputChannel.appendLine(msgRemoteBookDirectoryError);
+					return reject(err);
+				}
+
+				let options = {
+					headers: {
+						'User-Agent': 'request'
+					},
+					timeout: 800
+				};
+				let downloadRequest = request.get(this._assetUrl.href, options)
+					.on('error', (downloadError) => {
+						this.outputChannel.appendLine(msgRemoteBookDownloadError);
+						reject(downloadError);
+					})
+					.on('response', (response) => {
+						if (response.statusCode !== 200) {
+							this.outputChannel.appendLine(msgRemoteBookDownloadError);
+							return reject(response.statusMessage);
+						}
+					});
+				let remoteBookFullPath = new URL(this._local_path.href.concat('.zip'));
+				downloadRequest.pipe(fs.createWriteStream(remoteBookFullPath.href))
+					.on('close', async () => {
+						resolve(this.extractFiles(remoteBookFullPath));
+					})
+					.on('error', (downloadError) => {
+						this.outputChannel.appendLine(msgRemoteBookDownloadError);
+						reject(downloadError);
+						downloadRequest.abort();
+					});
+			});
+		});
+	}
+
+	public extractFiles(remoteBookFullPath: URL): void {
+		if (utils.getOSPlatform() === utils.Platform.Windows || utils.getOSPlatform() === utils.Platform.Mac) {
+			try {
+				let zippedFile = new zip(remoteBookFullPath.href);
+				zippedFile.extractAllTo(this._local_path.href);
+			} catch (err) {
+				this.outputChannel.appendLine(msgRemoteBookUnpackingError);
+			}
+			// Delete zip file
+			fs.unlink(remoteBookFullPath.href, (err) => {
+				if (err) {
+					this.outputChannel.appendLine(msgRemoteBookUnpackingError);
+				}
+			});
+			this.outputChannel.appendLine(msgRemoteBookDownloadComplete);
+
+		} else {
+			tar.extract({ file: remoteBookFullPath.href, cwd: this._local_path.href }).then(() => {
```

Comment:
Ideally this function should be async and then you should be awaiting these calls instead of chaining .then's